tests/test_interpolation.py

```diff
@@ -753,3 +756,250 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
     OmegaConf.register_new_resolver("test", lambda x: x)
     cfg = OmegaConf.create({"x": "${test:'${missing}'}", "missing": None})
     assert cfg.x == "None"
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key", "expected_value", "expected_node_type"),
+    [
+        pytest.param(
+            User(name="Bond", age=SI("${cast:int,'7'}")),
+            "age",
+            7,
+            IntegerNode,
+            id="expected_type",
+        ),
+        pytest.param(
+            # This example specifically test the case where intermediate resolver results
+            # cannot be cast to the same type as the key.
+            User(name="Bond", age=SI("${cast:int,${drop_last:${drop_last:7xx}}}")),
+            "age",
+            7,
+            IntegerNode,
+            id="intermediate_type_mismatch_ok",
+        ),
+        pytest.param(
+            # This example relies on the automatic casting of a string to int when
+            # assigned to an IntegerNode.
+            User(name="Bond", age=SI("${cast:str,'7'}")),
+            "age",
+            7,
+            IntegerNode,
+            id="convert_str_to_int",
+        ),
+        pytest.param(
+            MissingList(list=SI("${identity:[a, b, c]}")),
+            "list",
+            ["a", "b", "c"],
+            ListConfig,
+            id="list_str",
+        ),
+        pytest.param(
+            MissingList(list=SI("${identity:[0, 1, 2]}")),
+            "list",
+            ["0", "1", "2"],
+            ListConfig,
+            id="list_int_to_str",
+        ),
+        pytest.param(
+            MissingDict(dict=SI("${identity:{key1: val1, key2: val2}}")),
+            "dict",
+            {"key1": "val1", "key2": "val2"},
+            DictConfig,
+            id="dict_str",
+        ),
+        pytest.param(
+            MissingDict(dict=SI("${identity:{a: 0, b: 1}}")),
+            "dict",
+            {"a": "0", "b": "1"},
+            DictConfig,
+            id="dict_int_to_str",
+        ),
+    ],
+)
+def test_interpolation_type_validated_ok(
+    cfg: Any,
+    key: str,
+    expected_value: Any,
+    expected_node_type: Any,
+    register_identity: Any,
+) -> Any:
+    def cast(t: Any, v: Any) -> Any:
+        return {"str": str, "int": int}[t](v)  # cast `v` to type `t`
```

Comment:
cast is also a repeating function we use in many places. extract it as well?

Reply:
To keep tests simple I combined both `identity` and `cast` into the same `common_resolvers` feature

Reply:
sounds good.
You should also delete this inline cast function, right?

Reply:
Oops, you're right, thanks, I had missed it! Done in 57ee921